test/test_runlib.py

```diff
@@ -22,24 +22,126 @@
 import unittest
 import shutil
 import tempfile
+from in_toto import ssl_crypto
 from in_toto.runlib import in_toto_record_start, in_toto_record_stop, \
-    UNFINISHED_FILENAME_FORMAT
+    UNFINISHED_FILENAME_FORMAT, record_artifacts_as_dict, \
+    _apply_exclude_patterns
 from in_toto.util import generate_and_write_rsa_keypair, \
     prompt_import_rsa_key_from_file
 from in_toto.models.link import Link
 from in_toto.exceptions import SignatureVerificationError
+from simple_settings import settings
 
 
 WORKING_DIR = os.getcwd()
 
+class Test_ApplyExcludePatterns(unittest.TestCase):
+  """Test _apply_exclude_patterns(names, exclude_patterns) """
+
+  def test_apply_exclude_explict(self):
+    names = ["foo", "bar", "baz"]
+    patterns = ["foo", "bar"]
+    expected = ["baz"]
+    result = _apply_exclude_patterns(names, patterns)
+    self.assertListEqual(sorted(result), sorted(expected))
+
+  def test_apply_exclude_all(self):
+    names = ["foo", "bar", "baz"]
+    patterns = ["*"]
+    expected = []
+    result = _apply_exclude_patterns(names, patterns)
+    self.assertListEqual(sorted(result), sorted(expected))
+
+  def test_apply_exclude_multiple_star(self):
+    names = ["foo", "bar", "baz"]
+    patterns = ["*a*"]
+    expected = ["foo"]
+    result = _apply_exclude_patterns(names, patterns)
+    self.assertListEqual(sorted(result), sorted(expected))
+
+
+class TestRecordArtifactsAsDict(unittest.TestCase):
+  """Test record_artifacts_as_dict(artifacts). """
+
+  @classmethod
+  def setUpClass(self):
+    """Create and change into temp test directory with dummy artifacts. """
+    # Clear user set excludes
+    settings.ARTIFACT_EXCLUDES = []
+
+    self.test_dir = tempfile.mkdtemp()
+    os.chdir(self.test_dir)
+    open("foo", "w").write("foo")
+    open("bar", "w").write("bar")
+
+    os.mkdir("subdir")
+    os.mkdir("subdir/subsubdir")
+    open("subdir/foosub", "w").write("foosub")
+    open("subdir/subsubdir/foosubsub", "w").write("foosubsub")
```

Comment:
Yes and no. This TestCase only checks `runlib.record_artifacts_as_dict` which recursively traverses the passed list of paths and stores each file together with its hash. It does not care for hash collisions.

The "files with the same hash" corner case only appears in `verifylib.verify_match_rule`, where we try to map files from one artifact dict (as recorded above) to the files from another artifact dict.

Moreover, the hash collision issue has been dealt with in: https://github.com/in-toto/in-toto/commit/75ca2ce6f4b04fb9ca3bda66b282371883de9eb1
which I used for the [in-toto in Seattle](https://github.com/in-toto/in-toto/tree/seattle) show case, but which has not been PRed yet, because of ongoing discussions in https://github.com/in-toto/in-toto/issues/43.